### PR TITLE
9ipfs: add a IPFS 9P file server

### PR DESCRIPTION
**Pull request by fhs**
This is a simple read-only file server to access IPFS files though 9P.
Once mounted at mtpt, access the IPFS file/directory at mtpt/CID, where
CID is the IPFS content identifier.

Signed-off-by: Fazlul Shahriar <fshahriar@gmail.com>

### CODE REVIEW

**Review by rminnich — DISMISSED** [review]
very interesting. If you ever make this read/write I will finish up the # device to connect to it.

**Comment by rminnich** [comment]
this is really really cool, the more I think of it. Can you put some handy dandy instructions some where to show people how to use it ?
I could imagine have a harvey *IN SPI FLASH* that mounts ipfs, ... the dream.

And if we had a harvey that used only u-root as its in-flash root file system, this dream is easily achieved; with 16M flash parts and u-root, it's what we do with linux TODAY in google data centers, and other big companies do as well.

**Comment by floren** [comment]
Rather than specifying a root fileserver IP, you just give a hash. Seems
nice!

**Comment by gmacd** [comment]
I think this is how we can bind packages (e.g. go) - just mount via ipfs.  Will be much easier than building it individually.

**Comment by fhs** [comment]
> Can you put some handy dandy instructions some where to show people how to use it ?

Just run `9ipfs` in a rio window. You can use the `-debug` flag to see 9P trace. Then in a separate window:
```
cpu% mount /srv/9ipfs /n/ipfs
cpu% cd /n/ipfs/QmQPeNsJPyVWPFDVHb77w8G42Fvo15z4bG2X8D2GhfbSXc
cpu% ls -l
--r-xr-xr-x M 61 ipfs ipfs 1681 Feb  7  2006 about
--r-xr-xr-x M 61 ipfs ipfs  189 Feb  7  2006 contact
--r-xr-xr-x M 61 ipfs ipfs  311 Feb  7  2006 help
--r-xr-xr-x M 61 ipfs ipfs    4 Feb  7  2006 ping
--r-xr-xr-x M 61 ipfs ipfs 1681 Feb  7  2006 quick-start
--r-xr-xr-x M 61 ipfs ipfs 1091 Feb  7  2006 readme
--r-xr-xr-x M 61 ipfs ipfs 1162 Feb  7  2006 security-notes
cpu% cat readme
Hello and Welcome to IPFS!
...
```
This particular IPFS directory should be widely available, since it's used in the IPFS tutorial.

**Review by gmacd — APPROVED** [review]

File: cmd/9ipfs/go.mod [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module harvey-os.org/cmd/9ipfs
```
Comment (floren) [inline cmd/9ipfs/go.mod:1]: We should decide if we want to have go.mod files allowed per command, or if this should be unified with the top-level go.mod. In my professional work, we just maintain a single top-level go.mod per repo, but there may well be good reason to do it like this.

Comment (fhs) [reply cmd/9ipfs/go.mod:1]: ipfs has a lot of dependencies:
```
$ wc -l go.sum cmd/9ipfs/go.sum 
   101 go.sum
  1354 cmd/9ipfs/go.sum
```
so I wanted to keep it separate. Sub-modules are standard practice: see for example [gopls](https://github.com/golang/tools/blob/master/gopls/go.mod).

**Review by rminnich — APPROVED** [review]

**Comment by rminnich** [comment]
we need automerge but rest assured I'm merging this in a few seconds.

**Pull request merged**